pages/slice/language-reference/slice-grammar.md

```diff
@@ -2,14 +2,572 @@
+Hence the lexer always gives priority to producing keyword tokens, even when it could produce an identifier token.
+To use a keyword as an identifier, you can escape the keyword by prefixing it with a backslash.
+For scoped identifiers, each individual identifier must be escaped this way.
+It is legal to escape an identifier that is not a keyword, though doing so has no effect.
+
+```slice
+struct   // The struct keyword
+\struct  // An unscoped identifier
+\foo     // An unscoped identifier
+
+::\struct::\module // A globally scoped identifier
+```
+
+### Integer literals
+
+Slice supports three forms of integer literals:
+- A decimal literal for base-10 numbers. These consist of one or more digits between `0` and `9`.
+- A hexadecimal literal for base-16 numbers. These start with the prefix `0x`, followed by one or more digits between `0` and `9` or letters between `a` and `f`. Hexadecimal literals are not case-sensitive.
+- A binary literal for base-2 numbers. These start with the prefix `0b`, followed by one or more digits that are either `0` or `1`.
+
+Additionally, underscores can appear between any two characters of an integer literal.
+These underscores carry no semantic meaning and are discarded by the parser.
+
+Examples of valid integer literals:
+
+```slice
+123
+0xFF
+0b1010
+
+335_445_996
+0x_ab_cd_ef  // Same as 0xabcded
+0b0_______1  // Same as 0b01
+
+0x_0b1101  // hexadecimal literal of value 725249
+
+// All have a value of 0
+0
+0_00_0
+0b000
+0x0
+```
+
+{% callout type="note" %}
+Note that `-` signs are counted as separate tokens, and not as part of an integer literal.
+{% /callout %}
+
+The Slice grammar has 2 separate rules for unsigned and (possibly) signed integers.
+Slice does not support using a `+` sign to indicate an integer is positive, since all integers are positive by default.
+
+```ebnf {% showTitle=false %}
+Integer
+    : integer_literal
+    ;
+
+SignedInteger
+    : Integer
+    | "-" Integer
+    ;
+```
+
+### String literals
+
+String literals consist of any number of UTF8 characters, wrapped in a pair of double quotes.
+Any characters within a string literal are treated as opaque text by the parser.
+
+Slice supports escaping special characters within string literals, this allows you to include unprintable characters or literal double quote characters in your string. Characters can be escaped by prefixing them with a backslash.
+To write a literal backslash, you must escape it with another backslash: `\\`.
+
+When parsing, Slice will resolve any escaped characters, removing the `\` escape character where present.
+Apart from this, Slice performs no additional processing of string literals.
+
+Examples of valid string literals:
+
+```slice
+""
+"hello world"
+"the variable \"foo\" doesn't exist"
+"a backslash: '\\'"
+```
+
+### Attributes
+
+Slice supports two forms of attributes:
+- Local attributes: these consist of an attribute wrapped in a pair of single brackets.
```

Comment:
My preference is to use the same terminology as on https://docs.icerpc.dev/slice2/language-guide/attributes, namely attributes and file-level attributes. No local attribute / file attribute.

Reply:
I didn't know we were just calling them 'attributes' now.

I've always found 'local' and 'global' attributes to be bad terms, but,
what would be the umbrella term for "attributes" and "file-level attributes" though?...
Writing "Slice supports two forms of attributes: attributes and file-level attributes" would be odd.

Reply:
> what would be the umbrella term for "attributes" and "file-level attributes" though

Easy: we don't need an umbrella term. We have attributes and file attributes.

Reply:
Okay, just to make sure we agree on the concepts here, you're saying:

we only have attributes, and those attributes can on things.
`oneway` is an operation attribute, `cs::namespace` is a module attribute, and `cpp::header-ext` is a file attribute.

Slice just happens to require a slightly different syntax for file attributes? But it's still an _attribute_.

Reply:
I rewrote the whole attribute section.
We don't talk about local attributes anymore, we just talk about attributes.

And to apply an attribute to a file, use `[[`.
To apply an attribute to a definition/declaration, use `[`.
But they're all just _attributes_.